# **ENEM 2025 Mathematics Performance Analysis**

## A NumPy-Based Exploratory Data Analysis

This project explores Mathematics performance in the 2025 Brazilian National High School Exam (ENEM) using publicly available microdata provided by the Instituto Nacional de Estudos e Pesquisas Educacionais Anísio Teixeira (INEP).

The main goal is to investigate patterns in Mathematics performance while applying fundamental NumPy concepts to a real-world educational dataset.

## 1. Project Overview

### Objectives

This project aims to:

- explore Mathematics performance in ENEM 2025;
- investigate differences in Mathematics performance across geographical and school-related characteristics;
- apply NumPy techniques to a large real-world dataset;
- practice NumPy arrays, aggregation, indexing, slicing, sorting, and boolean filtering.

### Research Questions

This project investigates the following questions:

- How are Mathematics scores distributed in the ENEM 2025?
- How does mean Mathematics performance vary across exam states?
- How does mean Mathematics performance differ across school administrative dependency categories?
- How does mean Mathematics performance differ between urban and rural schools?
- Which exam municipalities present the highest mean Mathematics scores?

## 2. Data Source

The data used in this project come from the official ENEM 2025 microdata released by the Instituto Nacional de Estudos e Pesquisas Educacionais Anísio Teixeira (INEP).

The microdata contain anonymized information about ENEM participants and examination results. Official documentation provided by INEP is used throughout the project to interpret the variables and results.

## 3. Understanding the Data

### 3.1 Connecting to Google Drive

The original ENEM 2025 dataset is stored in Google Drive due to its large file size. Google Drive is mounted to the Colab environment to provide access to this file.

In [1]:
from google.colab import drive

drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
# Define path to the results dataset
results_path = "/content/drive/MyDrive/Data_Science_Portfolio/NumPy/ENEM_2025_Analysis/data/RESULTADOS_2025.csv"

### 3.2 Dataset Structure

Before loading the complete dataset into memory, its basic structure is inspected to identify the available variables and understand how the examination result data are organized.

In [3]:
# Open the results dataset
results = open(results_path, encoding="latin-1")

# Read the header
header = results.readline()

# Split the header into individual column names
results_columns = header.strip().split(";")

# Inspect the dataset columns
print(len(results_columns))
print(results_columns[:10])
print(results_columns[-10:])

# Display the first five observations
for _ in range(5):
  print(results.readline())

# Display column names and their indices
for index, column in enumerate(results_columns):
    print(index, column)

# Close the file
results.close()

70
['NU_SEQUENCIAL', 'NU_ANO', 'CO_ESCOLA', 'CO_MUNICIPIO_ESC', 'NO_MUNICIPIO_ESC', 'CO_UF_ESC', 'SG_UF_ESC', 'TP_DEPENDENCIA_ADM_ESC', 'TP_LOCALIZACAO_ESC', 'TP_SIT_FUNC_ESC']
['NU_NOTA_COMP3_AV3', 'NU_NOTA_COMP4_AV3', 'NU_NOTA_COMP5_AV3', 'TP_STATUS_REDACAO_AV4', 'NU_NOTA_AV4', 'NU_NOTA_COMP1_AV4', 'NU_NOTA_COMP2_AV4', 'NU_NOTA_COMP3_AV4', 'NU_NOTA_COMP4_AV4', 'NU_NOTA_COMP5_AV4']
206403;2025;33083592;3304557;Rio de Janeiro;33;RJ;2;1;1;3304557;Rio de Janeiro;33;RJ;1;1;1;1;1484;1447;1459;1472;524.1;624.2;534.9;653.6;ABDDCCAECABEEABBBADCAEEDDBDCBDBDCBBAADBDAECBB;EEDCDACBEECADBCAEBBCCEEBEADADBCABACBCDDACBCEB;DCCEADEDCEDBCCACCCBBDCABEBABBCCBCACABBCCBAEDC;CACBAEDBEBCBDECDDDDADDBCADBADBDBCBCBDEABCBBBC;0;CBDDCDAEAAAECBBCEAEEBDEEXABBBDXCDCBBBDCEADDCD;EDECADCEEDCADBDEBBACDDECBADABBCABACEEDDBCBCEA;DDDEABADDCECDCEBCECAECAEBCBCBEBABABDEBCEDABCABADDC;CECCADDBEBCACECBDDDBADEEABBECCDABEADDEBEAAXCC;1;120;160;140;120;100;640;1;600;120;120;160;120;80;1;680;120;200;120;120;120;;;;;;;;;;;;;;

3604651;20

### 3.3 Variable Selection

#### 3.3.1 Mathematics Variables

The following variables are directly related to the Mathematics examination:

In [4]:
# Select Mathematics-related variables

math_variables = (
    results_columns[17:18]
    + results_columns[21:22]
    + results_columns[25:26]
    + results_columns[29:30]
    + results_columns[34:35]
)

print(math_variables)

['TP_PRESENCA_MT', 'CO_PROVA_MT', 'NU_NOTA_MT', 'TX_RESPOSTAS_MT', 'TX_GABARITO_MT']


#### 3.3.2 Geographical Variables

Geographical variables were selected to investigate how Mathematics performance varies across different locations in Brazil.

In [5]:
geographical_variables = results_columns[10:14]
print(geographical_variables)

['CO_MUNICIPIO_PROVA', 'NO_MUNICIPIO_PROVA', 'CO_UF_PROVA', 'SG_UF_PROVA']


#### 3.3.3 School Variables

School-related variables were selected to investigate Mathematics performance according to the available characteristics and geographical information of the associated schools.

In [6]:
school_variables = results_columns[2:10]
print(school_variables)

['CO_ESCOLA', 'CO_MUNICIPIO_ESC', 'NO_MUNICIPIO_ESC', 'CO_UF_ESC', 'SG_UF_ESC', 'TP_DEPENDENCIA_ADM_ESC', 'TP_LOCALIZACAO_ESC', 'TP_SIT_FUNC_ESC']


#### 3.3.4 Selected Variables
The selected Mathematics, geographical, and school-related variables are combined into a single list.

In [7]:
selected_variables = math_variables + geographical_variables + school_variables

print(selected_variables)
print(f"Number of selected variables: {len(selected_variables)}")

['TP_PRESENCA_MT', 'CO_PROVA_MT', 'NU_NOTA_MT', 'TX_RESPOSTAS_MT', 'TX_GABARITO_MT', 'CO_MUNICIPIO_PROVA', 'NO_MUNICIPIO_PROVA', 'CO_UF_PROVA', 'SG_UF_PROVA', 'CO_ESCOLA', 'CO_MUNICIPIO_ESC', 'NO_MUNICIPIO_ESC', 'CO_UF_ESC', 'SG_UF_ESC', 'TP_DEPENDENCIA_ADM_ESC', 'TP_LOCALIZACAO_ESC', 'TP_SIT_FUNC_ESC']
Number of selected variables: 17


#### 3.4 Dataset Size

Before proceeding with data preparation, the total number of observations in the results dataset is calculated to provide context about the size of the data being analyzed.

In [8]:
# Count the number of observations
with open(results_path, encoding="latin-1") as file:
    num_rows = sum(1 for _ in file) - 1

print(f"Number of observations: {num_rows:,}")
print(f"Number of variables: {len(results_columns)}")

Number of observations: 4,810,772
Number of variables: 70


The results dataset contains **4,810,772 observations and 70 variables**. Given the size of the dataset, the analysis will focus only on the variables relevant to Mathematics performance, reducing unnecessary memory usage and processing.

## 4. Data Preparation

After exploring the structure of the dataset and identifying the variables relevant to the analysis, the next step is to prepare the data for numerical analysis.

Since the dataset contains more than 4.8 million observations, only the selected variables will be loaded in order to reduce memory usage and avoid unnecessary processing.

In [9]:
import numpy as np

### 4.1 Classifying the Selected Variables

Before loading the data into NumPy arrays, the selected variables are classified according to the type of information they represent. This distinction is important for determining how each variable should be stored and analyzed.

In [10]:
# Identify the column indices of the selected variables
selected_indices = []

for index, column in enumerate(results_columns):
    if column in selected_variables:
        selected_indices.append(index)

print(selected_indices)

[2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 17, 21, 25, 29, 34]


In [11]:
# Display the selected variables and their values for the first observation

with open(results_path, encoding="latin-1") as file:
    file.readline()  # Skip the header
    row = file.readline().strip().split(";")

for index in selected_indices:
    print(results_columns[index], ":", row[index])

CO_ESCOLA : 33083592
CO_MUNICIPIO_ESC : 3304557
NO_MUNICIPIO_ESC : Rio de Janeiro
CO_UF_ESC : 33
SG_UF_ESC : RJ
TP_DEPENDENCIA_ADM_ESC : 2
TP_LOCALIZACAO_ESC : 1
TP_SIT_FUNC_ESC : 1
CO_MUNICIPIO_PROVA : 3304557
NO_MUNICIPIO_PROVA : Rio de Janeiro
CO_UF_PROVA : 33
SG_UF_PROVA : RJ
TP_PRESENCA_MT : 1
CO_PROVA_MT : 1472
NU_NOTA_MT : 653.6
TX_RESPOSTAS_MT : CACBAEDBEBCBDECDDDDADDBCADBADBDBCBCBDEABCBBBC
TX_GABARITO_MT : CECCADDBEBCACECBDDDBADEEABBECCDABEADDEBEAAXCC


In [12]:
# Create lists for each variable type and classify the selected variables according to their data type
quantitative_variables = [
    "NU_NOTA_MT"
]

numeric_categorical_variables = [
    "TP_DEPENDENCIA_ADM_ESC",
    "TP_LOCALIZACAO_ESC",
    "TP_SIT_FUNC_ESC",
    "TP_PRESENCA_MT"
]

text_categorical_variables = [
    "NO_MUNICIPIO_ESC",
    "SG_UF_ESC",
    "NO_MUNICIPIO_PROVA",
    "SG_UF_PROVA"
]

identifier_variables = [
    "CO_ESCOLA",
    "CO_MUNICIPIO_ESC",
    "CO_UF_ESC",
    "CO_MUNICIPIO_PROVA",
    "CO_UF_PROVA",
    "CO_PROVA_MT"
]

text_sequence_variables = [
    "TX_RESPOSTAS_MT",
    "TX_GABARITO_MT"
]

### 4.2 Loading the Data into NumPy Arrays

The variables required for each stage of the analysis are loaded into NumPy arrays according to their data types. The Mathematics score (`NU_NOTA_MT`) is loaded first as the main quantitative variable of the analysis.

In [13]:
# Load Mathematics scores into a NumPy array
math_scores = np.genfromtxt(
    results_path,
    delimiter=";",
    skip_header=1,
    usecols=25,
    encoding="latin-1"
)

In [14]:
# Inspect the Mathematics scores array
print("Type:", type(math_scores))
print("Dimensions:", math_scores.ndim)
print("Shape:", math_scores.shape)
print("Size:", math_scores.size)
print("Data type:", math_scores.dtype)

Type: <class 'numpy.ndarray'>
Dimensions: 1
Shape: (4810772,)
Size: 4810772
Data type: float64


In [15]:
# Count missing Mathematics scores
missing_scores = np.isnan(math_scores).sum()

print("Missing scores:", missing_scores)

Missing scores: 1550436


In [16]:
# Remove missing Mathematics scores using boolean indexing

valid_scores_mask = ~np.isnan(math_scores)
valid_math_scores = math_scores[valid_scores_mask]

print("Valid scores:", valid_math_scores.size)

Valid scores: 3260336


In [17]:
# Calculate descriptive statistics for Mathematics scores
print(f"Minimum score: {valid_math_scores.min():.2f}")
print(f"Maximum score: {valid_math_scores.max():.2f}")
print(f"Mean score: {valid_math_scores.mean():.2f}")
print(f"Standard deviation: {valid_math_scores.std():.2f}")
print(f"Median score: {np.median(valid_math_scores):.2f}")

Minimum score: 0.00
Maximum score: 980.30
Mean score: 519.98
Standard deviation: 127.64
Median score: 500.00


In [18]:
# Investigate Mathematics scores equal to zero using boolean indexing

# Create a boolean mask
zero_scores_mask = valid_math_scores == 0

# Select the scores that satisfy the condition
zero_scores = valid_math_scores[zero_scores_mask]

# Count the number of zero scores
zero_scores_count = zero_scores.size

print("Scores equal to zero:", zero_scores_count)

Scores equal to zero: 893


**Observation:** Among the 3,260,336 valid Mathematics scores, 893 were equal to zero, representing approximately 0.027% of the valid scores.

In [19]:
# Identify participants with the maximum Mathematics score using boolean indexing

# Get the maximum score
maximum_score = valid_math_scores.max()

# Create a boolean mask
maximum_score_mask = valid_math_scores == maximum_score

# Select the maximum scores
maximum_scores = valid_math_scores[maximum_score_mask]

# Count the number of maximum scores
maximum_score_count = maximum_scores.size

print(f"Maximum score: {maximum_score:.2f}")
print("Number of maximum scores:", maximum_score_count)

Maximum score: 980.30
Number of maximum scores: 88


**Observation:** The highest Mathematics score in ENEM 2025 was 980.30, achieved by 88 participants.

In [20]:
# Calculate the quartiles of Mathematics scores

quartiles = np.percentile(valid_math_scores, [25, 50, 75])

print(quartiles)

[416.5 500.  606.8]


**Observation:** The middle 50% of valid Mathematics scores ranged from 416.5 to 606.8, with a median score of 500.0.

In [21]:
# Get the total number of valid Mathematics scores
total = valid_math_scores.size

# Scores from 0 to 100
scores_0to100_mask = (valid_math_scores >= 0) & (valid_math_scores < 100)
number0to100 = (valid_math_scores[scores_0to100_mask]).size
percentage0to100 = (number0to100 / total) * 100
print(f"{percentage0to100:.2f}%")

# Scores from 100 to 200
scores_100to200_mask = (valid_math_scores >= 100) & (valid_math_scores < 200)
number100to200 = (valid_math_scores[scores_100to200_mask]).size
percentage100to200 = (number100to200 / total) * 100
print(f"{percentage100to200:.2f}%")

0.03%
0.00%


**Observation:** The first score range, from 0 to 100, contains 893 observations. Since this is exactly the same number of scores equal to zero, there are no positive scores below 100. Additionally, no scores were found in the 100 to 200 range. This raises a new question: **what is the lowest positive Mathematics score in the dataset?**

In [22]:
# Find the lowest positive Mathematics score
scores_higher_than_0_mask = valid_math_scores > 0
print(valid_math_scores[scores_higher_than_0_mask].min())


312.6


**Finding:** The lowest positive Mathematics score in the dataset is 312.6. Therefore, no valid Mathematics scores were observed between 0 and 312.6.

In [23]:
# Scores from 300 to 400
scores_300to400_mask = (valid_math_scores >= 300) & (valid_math_scores < 400)
number300to400 = (valid_math_scores[scores_300to400_mask]).size
percentage300to400 = (number300to400 / total) * 100
print(f"{percentage300to400:.2f}%")

# Scores from 400 to 500
scores_400to500_mask = (valid_math_scores >= 400) & (valid_math_scores < 500)
number400to500 = (valid_math_scores[scores_400to500_mask]).size
percentage400to500 = (number400to500 / total) * 100
print(f"{percentage400to500:.2f}%")

# Scores from 500 to 600
scores_500to600_mask = (valid_math_scores >= 500) & (valid_math_scores < 600)
number500to600 = (valid_math_scores[scores_500to600_mask]).size
percentage500to600 = (number500to600 / total) * 100
print(f"{percentage500to600:.2f}%")

# Scores from 600 to 700
scores_600to700_mask = (valid_math_scores >= 600) & (valid_math_scores < 700)
number600to700 = (valid_math_scores[scores_600to700_mask]).size
percentage600to700 = (number600to700 / total) * 100
print(f"{percentage600to700:.2f}%")

# Scores from 700 to 800
scores_700to800_mask = (valid_math_scores >= 700) & (valid_math_scores < 800)
number700to800 = (valid_math_scores[scores_700to800_mask]).size
percentage700to800 = (number700to800 / total) * 100
print(f"{percentage700to800:.2f}%")

# Scores from 800 to 900
scores_800to900_mask = (valid_math_scores >= 800) & (valid_math_scores < 900)
number800to900 = (valid_math_scores[scores_800to900_mask]).size
percentage800to900 = (number800to900 / total) * 100
print(f"{percentage800to900:.2f}%")

# Scores from 900 to the maximum score
scores_900to_max_mask = (valid_math_scores >= 900) & (valid_math_scores <= maximum_score)
number900to_max = valid_math_scores[scores_900to_max_mask].size
percentage900to_max = (number900to_max / total) * 100

print(f"{percentage900to_max:.2f}%")

19.65%
30.31%
23.65%
16.06%
7.81%
2.20%
0.29%


### 4.3 Mathematics Score Distribution by Range

The distribution of valid Mathematics scores across 100-point intervals is:

- **0–100:** 0.03%
- **100–200:** 0.00%
- **200–300:** 0.00%
- **300–400:** 19.65%
- **400–500:** 30.31%
- **500–600:** 23.65%
- **600–700:** 16.06%
- **700–800:** 7.81%
- **800–900:** 2.20%
- **900–980.30:** 0.29%

The largest concentration of scores is found between **400 and 500**, accounting for 30.31% of the valid observations. Together, the ranges from **400 to 600** contain more than half of all valid Mathematics scores.

The proportion of observations decreases considerably in the higher score ranges. While 16.06% of the scores fall between 600 and 700, this proportion decreases to 7.81% between 700 and 800, 2.20% between 800 and 900, and only 0.29% between 900 and the maximum observed score of 980.30.

The lower end of the distribution presents a particular pattern. The 0–100 range contains 893 observations (0.03%), all of which correspond to scores equal to zero. No scores were observed between 100 and 300, and the lowest positive Mathematics score in the dataset is **312.6**.

### 4.4 Loading Geographical Data

To analyze Mathematics performance across different locations, the state where each participant took the exam (`SG_UF_PROVA`) is loaded into a NumPy array. The order of the observations must be preserved so that each state remains aligned with its corresponding Mathematics score.

In [24]:
# Load the exam state (UF) into a NumPy array

test_uf = np.genfromtxt(
    results_path,
    delimiter=";",
    skip_header=1,
    usecols=13,
    dtype=str,
    encoding="latin-1"
)

In [25]:
# Inspect the exam state array

print("Dimensions:", test_uf.ndim)
print("Shape:", test_uf.shape)
print("Size:", test_uf.size)
print("Data type:", test_uf.dtype)

Dimensions: 1
Shape: (4810772,)
Size: 4810772
Data type: <U2


In [26]:
# Check whether the state and Mathematics score arrays have the same size

test_uf.size == math_scores.size

True

In [27]:
# Keep only the states associated with valid Mathematics scores

valid_test_uf = test_uf[valid_scores_mask]

# Identify the unique states in the dataset

unique_ufs = np.unique(valid_test_uf)

print(unique_ufs)
print("Number of states:", unique_ufs.size)

['AC' 'AL' 'AM' 'AP' 'BA' 'CE' 'DF' 'ES' 'GO' 'MA' 'MG' 'MS' 'MT' 'PA'
 'PB' 'PE' 'PI' 'PR' 'RJ' 'RN' 'RO' 'RR' 'RS' 'SC' 'SE' 'SP' 'TO']
Number of states: 27


In [28]:
# Calculate the mean Mathematics score for each state

uf_means = []

for uf in unique_ufs:
    uf_mask = valid_test_uf == uf
    uf_scores = valid_math_scores[uf_mask]
    uf_mean = uf_scores.mean()

    uf_means.append(uf_mean)

# Convert the list of state means into a NumPy array

uf_means = np.array(uf_means)

# Get the indices that would sort the mean scores in ascending order

sorted_indices = np.argsort(uf_means)

# Sort states and mean scores using the same indices

sorted_ufs = unique_ufs[sorted_indices]
sorted_means = uf_means[sorted_indices]

# Display states ordered by mean Mathematics score

for i in range(sorted_ufs.size):
    print(f"{sorted_ufs[i]}: {sorted_means[i]:.2f}")

AM: 464.61
AP: 466.41
PA: 467.07
MA: 470.13
AC: 479.14
TO: 486.87
BA: 486.94
RO: 488.15
PI: 491.42
RR: 492.32
AL: 492.59
SE: 494.82
PB: 505.39
MT: 508.83
PE: 510.46
CE: 511.24
MS: 513.39
RN: 515.61
GO: 521.76
RJ: 537.61
PR: 544.66
ES: 545.10
DF: 547.98
MG: 548.20
RS: 549.08
SC: 553.49
SP: 560.34


### 4.5 Mathematics Performance by Exam State

The mean Mathematics score was calculated for each Brazilian state according to the state where the exam was taken (`SG_UF_PROVA`). The state averages were then sorted in ascending order to facilitate comparison.

The lowest mean score was observed in **Amazonas (AM)**, with **464.61 points**, followed by **Amapá (AP)** with 466.41 and **Pará (PA)** with 467.07.

At the other end of the distribution, **São Paulo (SP)** had the highest mean Mathematics score, with **560.34 points**, followed by **Santa Catarina (SC)** with 553.49 and **Rio Grande do Sul (RS)** with 549.08.

The difference between the highest and lowest state averages was **95.73 points**, indicating substantial variation in mean Mathematics performance across the states where the exam was taken.

It is important to note that `SG_UF_PROVA` represents the state where the participant took the exam and does not necessarily represent their state of residence or schooling.

### 4.6 School Administrative Dependency

The school administrative dependency variable (`TP_DEPENDENCIA_ADM_ESC`) is loaded to investigate differences in Mathematics performance according to the administrative category of the school.

In [29]:
# Load the school administrative dependency into a NumPy array

school_dependency = np.genfromtxt(
    results_path,
    delimiter=";",
    skip_header=1,
    usecols=7,
    encoding="latin-1"
)

In [30]:
# Check whether the school dependency and Mathematics score arrays have the same size

school_dependency.size == math_scores.size

True

In [31]:
# Keep only the school dependency values associated with valid Mathematics scores

valid_school_dependency = school_dependency[valid_scores_mask]

# Check whether the filtered arrays have the same size

print("Valid Mathematics scores:", valid_math_scores.size)
print("Corresponding school dependency values:", valid_school_dependency.size)

Valid Mathematics scores: 3260336
Corresponding school dependency values: 3260336


In [32]:
# Identify the unique school dependency values

unique_school_dependencies = np.unique(valid_school_dependency)

print(unique_school_dependencies)

[ 1.  2.  3.  4. nan]


In [33]:
# Create a mask for records with available school dependency information

school_info_mask = ~np.isnan(valid_school_dependency)

# Keep only records with available school dependency information

school_dependency_clean = valid_school_dependency[school_info_mask]
school_math_scores = valid_math_scores[school_info_mask]

print("School dependency records:", school_dependency_clean.size)
print("Corresponding Mathematics scores:", school_math_scores.size)

School dependency records: 1292603
Corresponding Mathematics scores: 1292603


In [34]:
# Identify the school dependency categories after removing missing values

clean_dependencies = np.unique(school_dependency_clean)

print(clean_dependencies)

[1. 2. 3. 4.]


After removing missing values, 1,292,603 observations remained with both a valid Mathematics score and available school administrative dependency information.

According to the ENEM 2025 Microdata Dictionary, the categories of `TP_DEPENDENCIA_ADM_ESC` are:

- **1:** Federal
- **2:** State
- **3:** Municipal
- **4:** Private

In [35]:
# Calculate the mean Mathematics score for each school dependency category

dependency_means = []

for dependency in clean_dependencies:
    dependency_mask = school_dependency_clean == dependency
    dependency_scores = school_math_scores[dependency_mask]
    dependency_mean = dependency_scores.mean()

    dependency_means.append(dependency_mean)

# Convert the list of dependency means into a NumPy array

dependency_means = np.array(dependency_means)

# Define the school dependency category labels

dependency_labels = np.array([
    "Federal",
    "State",
    "Municipal",
    "Private"
])

# Get the indices that would sort the mean scores in ascending order

sorted_indices = np.argsort(dependency_means)

# Sort school dependency categories and mean scores using the same indices

sorted_dependency_labels = dependency_labels[sorted_indices]
sorted_dependency_means = dependency_means[sorted_indices]

# Display school dependency categories ordered by mean Mathematics score

for i in range(sorted_dependency_labels.size):
    print(f"{sorted_dependency_labels[i]}: {sorted_dependency_means[i]:.2f}")

State: 480.63
Municipal: 532.26
Federal: 586.95
Private: 620.32


**Observation:** Mathematics performance varies substantially across school administrative dependency categories.

**State schools** presented the lowest mean Mathematics score, with **480.63 points**, followed by **municipal schools**, with **532.26 points**. Higher averages were observed among **federal schools**, with **586.95 points**, and **private schools**, which presented the highest mean score, at **620.32 points**.

The difference between the highest and lowest averages was **139.69 points**, with private schools averaging substantially higher Mathematics scores than state schools.

These results describe differences observed in the ENEM 2025 data and should not be interpreted as evidence that school administrative dependency itself causes differences in Mathematics performance.

### 4.7 School Location

The school location variable (`TP_LOCALIZACAO_ESC`) is loaded to investigate differences in Mathematics performance according to the location of the school.

In [36]:
# Load the school location into a NumPy array

school_location = np.genfromtxt(
    results_path,
    delimiter=";",
    skip_header=1,
    usecols=8,
    encoding="latin-1"
)

In [37]:
# Check whether the school location and Mathematics score arrays have the same size

school_location.size == math_scores.size

# Keep only the school location values associated with valid Mathematics scores

valid_school_location = school_location[valid_scores_mask]

# Check whether the filtered arrays have the same size

print("Valid Mathematics scores:", valid_math_scores.size)
print("Corresponding school location values:", valid_school_location.size)

# Identify the unique school location values

unique_school_locations = np.unique(valid_school_location)

print(unique_school_locations)

Valid Mathematics scores: 3260336
Corresponding school location values: 3260336
[ 1.  2. nan]


In [38]:
# Create a mask for records with available school location information

school_location_info_mask = ~np.isnan(valid_school_location)

# Keep only records with available school location information

school_location_clean = valid_school_location[school_location_info_mask]
school_location_math_scores = valid_math_scores[school_location_info_mask]

# Check whether the filtered arrays have the same size

print("School location records:", school_location_clean.size)
print("Corresponding Mathematics scores:", school_location_math_scores.size)

# Identify the school location categories after removing missing values

clean_school_locations = np.unique(school_location_clean)

print(clean_school_locations)

School location records: 1292603
Corresponding Mathematics scores: 1292603
[1. 2.]


After removing missing values, 1,292,603 observations remained with both a valid Mathematics score and available school location information.

According to the ENEM 2025 Microdata Dictionary, the categories of `TP_LOCALIZACAO_ESC` are:

- **1:** Urban
- **2:** Rural

Both school location categories are represented in the filtered dataset.

In [39]:
# Calculate the mean Mathematics score for each school location category

location_means = []

for location in clean_school_locations:
    location_mask = school_location_clean == location
    location_scores = school_location_math_scores[location_mask]
    location_mean = location_scores.mean()

    location_means.append(location_mean)

# Convert the list of location means into a NumPy array

location_means = np.array(location_means)

# Define the school location category labels

location_labels = np.array([
    "Urban",
    "Rural"
])

# Display the mean Mathematics score for each school location category

for i in range(location_labels.size):
    print(f"{location_labels[i]}: {location_means[i]:.2f}")

Urban: 516.27
Rural: 465.96


The mean Mathematics score was calculated for each school location category, considering only observations with both a valid Mathematics score and available school location information.

Urban schools presented a mean Mathematics score of **516.27 points**, while rural schools had a mean score of **465.96 points**.

The difference between the two categories was **50.31 points**, with observations associated with urban schools presenting a higher average Mathematics score.

These results indicate an association between school location and Mathematics performance in the analyzed data. However, the difference should not be interpreted as evidence that school location itself causes higher or lower performance, since other socioeconomic, educational, and regional factors may also be associated with the observed scores.


### 4.8 Mathematics Performance by Exam Municipality

The municipality where the exam was taken (`NO_MUNICIPIO_PROVA`) is analyzed to investigate differences in mean Mathematics performance across Brazilian municipalities.

The analysis aims to identify the **50 exam municipalities with the highest mean Mathematics scores** in the ENEM 2025.

In [40]:
# Load the exam municipality into a NumPy array

test_municipality = np.genfromtxt(
    results_path,
    delimiter=";",
    skip_header=1,
    usecols=11,
    dtype=str,
    encoding="latin-1"
)

In [41]:
# Check whether the municipality and Mathematics score arrays have the same size

test_municipality.size == math_scores.size

# Keep only the municipalities associated with valid Mathematics scores

valid_test_municipality = test_municipality[valid_scores_mask]

# Check whether the filtered arrays have the same size

print("Valid Mathematics scores:", valid_math_scores.size)
print("Corresponding municipalities:", valid_test_municipality.size)

Valid Mathematics scores: 3260336
Corresponding municipalities: 3260336


In [42]:
# Identify the unique exam municipalities

unique_municipalities = np.unique(valid_test_municipality)
print(f"Number of unique exam municipalities: {unique_municipalities.size}")

# Calculate the mean Mathematics score for each exam municipality

municipality_means = []

for municipality in unique_municipalities:
    municipality_mask = valid_test_municipality == municipality
    municipality_scores = valid_math_scores[municipality_mask]
    municipality_mean = municipality_scores.mean()

    municipality_means.append(municipality_mean)

# Convert the list of municipality means into a NumPy array
municipality_means = np.array(municipality_means)

Number of unique exam municipalities: 1768


In [43]:
# Identify the state corresponding to each exam municipality

municipality_ufs = []

for municipality in unique_municipalities:
    municipality_mask = valid_test_municipality == municipality
    municipality_uf = valid_test_uf[municipality_mask][0]

    municipality_ufs.append(municipality_uf)

# Convert the list of states into a NumPy array
municipality_ufs = np.array(municipality_ufs)

# Get the indices that sort municipality means in ascending order

sorted_indices = np.argsort(municipality_means)

# Sort municipalities, states, and their mean scores

sorted_municipalities = unique_municipalities[sorted_indices]
sorted_municipality_ufs = municipality_ufs[sorted_indices]
sorted_municipality_means = municipality_means[sorted_indices]


# Select the 50 municipalities with the highest mean scores

top_50_municipalities = sorted_municipalities[-50:]
top_50_ufs = sorted_municipality_ufs[-50:]
top_50_means = sorted_municipality_means[-50:]

In [44]:
# Display the top 50 municipalities in descending order

for i in range(49, -1, -1):
    rank = 50 - i

    print(
        f"{rank}. {top_50_municipalities[i]} - "
        f"{top_50_ufs[i]}: {top_50_means[i]:.2f}"
    )

1. Veranópolis - RS: 617.67
2. Lajeado - RS: 615.14
3. Valinhos - SP: 615.01
4. Ivoti - RS: 607.33
5. São José dos Campos - SP: 605.55
6. Estrela - RS: 602.30
7. Vinhedo - SP: 601.28
8. Jundiaí - SP: 596.66
9. Bento Gonçalves - RS: 595.40
10. Arroio do Meio - RS: 594.99
11. São Carlos - SP: 594.45
12. Americana - SP: 593.78
13. São Caetano do Sul - SP: 593.73
14. Carlos Barbosa - RS: 592.86
15. Itaúna - MG: 591.65
16. Formiga - MG: 591.38
17. São João da Boa Vista - SP: 589.79
18. Horizontina - RS: 588.71
19. Itatiba - SP: 588.40
20. Belo Horizonte - MG: 587.64
21. Vitória - ES: 587.18
22. Jaraguá do Sul - SC: 586.32
23. Divinópolis - MG: 585.78
24. Niterói - RJ: 585.50
25. Florianópolis - SC: 585.09
26. Amparo - SP: 585.09
27. Guaratinguetá - SP: 584.67
28. Timóteo - MG: 584.43
29. Campinas - SP: 584.40
30. Indaiatuba - SP: 583.98
31. Botucatu - SP: 583.98
32. Vargem Grande do Sul - SP: 583.52
33. Nova Petrópolis - RS: 583.07
34. Paulínia - SP: 582.66
35. Araras - SP: 582.60
36. Jagua

**Observation:** Among the 1,768 exam municipalities represented in the valid Mathematics score data, **Veranópolis (RS)** presented the highest mean score, with **617.67 points**, followed by **Lajeado (RS)** with **615.14 points** and **Valinhos (SP)** with **615.01 points**.

The ranking refers to the municipality where the exam was taken and does not necessarily represent the participant's municipality of residence or schooling.

## 5. Conclusion

This project explored Mathematics performance in the ENEM 2025 using official INEP microdata and NumPy as the main data analysis tool. The analysis included more than 4.8 million records and focused on valid Mathematics scores, examining their overall distribution as well as differences across geographical and school-related categories.

The results showed substantial variation in Mathematics performance. Differences were observed across exam states, school administrative dependencies, and school locations. Private and federal schools presented higher mean scores than state and municipal schools, while urban schools had a higher mean score than rural schools. At the municipal level, the analysis also identified the 50 exam municipalities with the highest average Mathematics scores. These comparisons are descriptive and should not be interpreted as evidence of causal relationships, since performance may be associated with several socioeconomic, educational, and regional factors.

A limitation of this analysis is the absence of participant-level socioeconomic information linked to the Mathematics results in the dataset used. Therefore, the observed differences could not be examined in relation to factors such as socioeconomic and demographic characteristics. Future analyses using datasets that allow these variables to be linked to individual performance could provide a more comprehensive understanding of the patterns observed in ENEM Mathematics scores.

From a technical perspective, the project provided practical experience with NumPy arrays, boolean masks, missing-value handling, indexing, filtering, descriptive statistics, `np.unique()`, `np.argsort()`, and the alignment of multiple arrays. Working with a large real-world dataset also demonstrated how these NumPy operations can be combined to transform raw data into structured and interpretable analyses.